# Aggregations: Min, Max, and Everything In Between

Often when faced with a large amount of data, a first step is to compute summary statistics for the data in question. Perhaps the most common summary statistics are the mean and standard deviation, which allow you to summarize the "typical" values in a dataset, but other aggregates are useful as well (the sum, product, median, minimum and maximum, quantiles, etc.).

NumPy has fast built-in aggregation functions for working on arrays; we'll discuss and demonstrate some of them here.

# Summing the Values in an Array
As a quick example, consider computing the sum of all values in an array. Python itself can do this using the built-in sum function:

In [1]:
import numpy as np


In [3]:
L=np.random.random(100)
L

array([0.28233582, 0.48236202, 0.20233101, 0.71845141, 0.42372554,
       0.39044503, 0.73130994, 0.6438406 , 0.72761422, 0.37530773,
       0.56252082, 0.92356555, 0.22163102, 0.18121119, 0.10918504,
       0.88525802, 0.28517876, 0.42459598, 0.45726162, 0.63116448,
       0.64088049, 0.37921412, 0.47020355, 0.15752702, 0.03010708,
       0.83404693, 0.16135508, 0.05829631, 0.03551256, 0.38784631,
       0.9643    , 0.34814468, 0.66073219, 0.71022669, 0.45106714,
       0.87656218, 0.54346154, 0.60569571, 0.02839696, 0.07294074,
       0.00584468, 0.71680077, 0.37711965, 0.09328622, 0.2010204 ,
       0.71853972, 0.25694185, 0.4983481 , 0.74480381, 0.47499453,
       0.64812826, 0.2662945 , 0.47479476, 0.35444887, 0.12380037,
       0.68479089, 0.56465371, 0.89872744, 0.90236331, 0.71316039,
       0.87427405, 0.8011151 , 0.28671805, 0.25728804, 0.46925873,
       0.19479993, 0.78712427, 0.28670765, 0.87258304, 0.02268225,
       0.1237391 , 0.9417368 , 0.55203248, 0.8692258 , 0.57218

In [4]:
sum(L)

49.028764813964635

The syntax is quite similar to that of NumPy's sum function, and the result is the same in the simplest case:

In [5]:
np.sum(L)

49.02876481396464

However, because it executes the operation in compiled code, NumPy's version of the operation is computed much more quickly:

In [6]:
big_array = np.random.rand(1000000)
%timeit sum(big_array)
%timeit np.sum(big_array)

85.5 ms ± 3.9 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
1.25 ms ± 44.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


e careful, though: the sum function and the np.sum function are not identical, which can sometimes lead to confusion! In particular, their optional arguments have different meanings, and np.sum is aware of multiple array dimensions, as we will see in the following section.

### Minimum and Maximum
Similarly, Python has built-in min and max functions, used to find the minimum value and maximum value of any given array:

In [7]:
min(big_array), max(big_array)

(6.347919390137235e-08, 0.999997937356572)

NumPy's corresponding functions have similar syntax, and again operate much more quickly:

In [9]:
np.min(big_array),np.max(big_array)

(6.347919390137235e-08, 0.999997937356572)

In [10]:
%timeit min(big_array)
%timeit np.min(big_array)

58.5 ms ± 1.73 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
548 µs ± 25 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


For min, max, sum, and several other NumPy aggregates, a shorter syntax is to use methods of the array object itself:

In [11]:
print(big_array.min(), big_array.max(), big_array.sum())

6.347919390137235e-08 0.999997937356572 500067.0287676032


Whenever possible, make sure that you are using the NumPy version of these aggregates when operating on NumPy arrays!

# Multi dimensional aggregates
One common type of aggregation operation is an aggregate along a row or column. Say you have some data stored in a two-dimensional array

In [12]:
M = np.random.random((3, 4))
print(M)

[[0.42259582 0.78297894 0.81821651 0.97882414]
 [0.76608267 0.05002797 0.51397585 0.11249486]
 [0.30143407 0.51360252 0.33459276 0.61697201]]


By default, each NumPy aggregation function will return the aggregate over the entire array

In [13]:
M.sum()

6.2117981113994105

In [14]:
np.sum(M)

6.2117981113994105

Aggregation functions take an additional argument specifying the axis along which the aggregate is computed. For example, we can find the minimum value within each column by specifying axis=0:

In [15]:
M.min(axis=0)

array([0.30143407, 0.05002797, 0.33459276, 0.11249486])

In [16]:
M.min(axis=1)

array([0.42259582, 0.05002797, 0.30143407])

In [17]:
M.sum(axis=0)

array([1.49011256, 1.34660943, 1.66678512, 1.708291  ])

In [18]:
M.sum(axis=1)

array([3.00261541, 1.44258134, 1.76660136])

The function returns four values, corresponding to the four columns of numbers.

Similarly, we can find the maximum value within each row:



In [19]:
M.max(axis=1)

array([0.97882414, 0.76608267, 0.61697201])

The way the axis is specified here can be confusing to users coming from other languages. The axis keyword specifies the dimension of the array that will be collapsed, rather than the dimension that will be returned. So specifying axis=0 means that the first axis will be collapsed: for two-dimensional arrays, this means that values within each column will be aggregated.

Example: What is the Average Height of US Presidents?
Aggregates available in NumPy can be extremely useful for summarizing a set of values. As a simple example, let's consider the heights of all US presidents. This data is available in the file president_heights.csv, which is a simple comma-separated list of labels and values:

In [21]:
!head -4 data/president_heights.csv

'head' is not recognized as an internal or external command,
operable program or batch file.


In [27]:
import pandas as pd
data = pd.read_csv(r'C:\Users\jibin\Desktop\har\NUMPY\New folder\csv\president_heights.csv')


In [28]:
data.head()

order               name  height(cm)
0      1  George Washington         189
1      2         John Adams         170
2      3   Thomas Jefferson         189
3      4      James Madison         163
4      5       James Monroe         183

In [29]:

heights = np.array(data['height(cm)'])
print(heights)

[189 170 189 163 183 171 185 168 173 183 173 173 175 178 183 193 178 173
 174 183 183 168 170 178 182 180 183 178 182 188 175 179 183 193 182 183
 177 185 188 188 182 185]


Now that we have this data array, we can compute a variety of summary statistics:

In [30]:
print("Mean height:       ", heights.mean())
print("Standard deviation:", heights.std())
print("Minimum height:    ", heights.min())
print("Maximum height:    ", heights.max())

Mean height:        179.73809523809524
Standard deviation: 6.931843442745892
Minimum height:     163
Maximum height:     193


Note that in each case, the aggregation operation reduced the entire array to a single summarizing value, which gives us information about the distribution of values. We may also wish to compute quantiles:

In [31]:
print("25th percentile:   ", np.percentile(heights, 25))
print("Median:            ", np.median(heights))
print("75th percentile:   ", np.percentile(heights, 75))

25th percentile:    174.25
Median:             182.0
75th percentile:    183.0


e see that the median height of US presidents is 182 cm, or just shy of six feet.

Of course, sometimes it's more useful to see a visual representation of this data, which we can accomplish using tools in Matplotlib (we'll discuss Matplotlib more fully in Chapter 4). For example, this code generates the following chart:

In [ ]:
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn; seaborn.set()  # set plot style